overlord/auth/auth.go

```diff
@@ -84,6 +86,51 @@ func User(st *state.State, id int) (*UserState, error) {
 	return nil, fmt.Errorf("invalid user")
 }
 
+// GetUserFromRequest extracts user information from request and return the respective user in state, if valid
+func GetUserFromRequest(st *state.State, req *http.Request) (*UserState, error) {
+	// extract macaroons data from request
+	header := req.Header.Get("Authorization")
+	if header == "" {
+		return nil, nil
+	}
+
+	const errorMsg = "unauthorized"
+	auth := strings.SplitN(header, " ", 2)
+	if len(auth) != 2 || auth[0] != "Macaroon" {
+		return nil, fmt.Errorf(errorMsg)
+	}
+
+	var macaroon string
+	var discharges []string
+	for _, field := range strings.Split(auth[1], ",") {
+		field := strings.TrimSpace(field)
+		if strings.HasPrefix(field, `root="`) {
+			macaroon = field[6 : len(field)-1]
+		}
+		if strings.HasPrefix(field, `discharge="`) {
+			discharges = append(discharges, field[11:len(field)-1])
+		}
+	}
+
+	if macaroon == "" || len(discharges) == 0 {
+		return nil, fmt.Errorf(errorMsg)
+	}
+
+	var authStateData AuthState
+	err := st.Get("auth", &authStateData)
+
+	if err != nil {
+		return nil, fmt.Errorf(errorMsg)
+	}
+
+	for _, user := range authStateData.Users {
+		if user.Macaroon == macaroon && reflect.DeepEqual(discharges, user.Discharges) {
```

Comment:
Let's please not use reflection here:
```
NextUser:
    for _, user := range authStateData.Users {
            if user.Macaroon != macaroon {
                    continue
            }
            for i, d := user.Discharges {
                    if d != discharges[i] {
                            continue NextUser
                    }
            }
            return nil
    }
    return fmt.Errorf("invalid authentication")
```

Also, the two discharges probably need to be sorted for that comparison.
`
sort.Strings(...)
`